### ISSUE TITLE
constructed plugin - keyed_groups cannot be used to create groups based on hostvars

### ISSUE THREAD

**Issue by christianpinger** [ansible/ansible#54476]
##### SUMMARY

When I try to create groups based on hostvars using the constructed plugin with keyed_groups, no hostvars are found.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

inventory-plugin: constructed

##### ANSIBLE VERSION

```paste below
ansible 2.7.9
  config file = /Users/charendt/git/ankibinn/ansible.cfg
  configured module search path = ['/Users/charendt/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.2 (default, Feb 12 2019, 08:15:36) [Clang 10.0.0 (clang-1000.11.45.5)]
```

##### CONFIGURATION

```paste below
HOST_KEY_CHECKING(/Users/charendt/git/ankibinn/ansible.cfg) = False
INVENTORY_ENABLED(/Users/charendt/git/ankibinn/ansible.cfg) = ['ini', 'constructed']
PERSISTENT_CONNECT_TIMEOUT(/Users/charendt/git/ankibinn/ansible.cfg) = 1800
RETRY_FILES_ENABLED(/Users/charendt/git/ankibinn/ansible.cfg) = False
```

##### OS / ENVIRONMENT

macOS, CLI, local, inventory parsing

##### STEPS TO REPRODUCE

I have two inventories. 

One ini inventory with the actual hosts called simpleinventory.ini:
```ini
host1
```

And one constructed inventory called inventory.config:
```yaml
---
plugin: constructed
strict: no

keyed_groups:
  - prefix: site
    key: mylocation
```

In addition I have hostvars specified in host_vars/host1/host1.yml
```yaml
---
mylocation: box1
```

##### EXPECTED RESULTS

Running `ansible-inventory -i simpleinventory.ini -i inventory.config --graph` will create groups based on the hostvar mylocation:

```
@all:
  |--@site_box1:
  |  |--host1
  |--@ungrouped:
  |  |-----
```

##### ACTUAL RESULTS

No groups are created

```paste below
ansible-inventory -i simpleinventory.ini -i inventory.config --graph -vvvv
ansible-inventory 2.7.9
  config file = /Users/charendt/git/ankibinn/ansible.cfg
  configured module search path = ['/Users/charendt/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible-inventory
  python version = 3.7.2 (default, Feb 12 2019, 08:15:36) [Clang 10.0.0 (clang-1000.11.45.5)]
Using /Users/charendt/git/ankibinn/ansible.cfg as config file
setting up inventory plugins
Parsed /Users/charendt/git/ankibinn/simpleinventory.ini inventory source with ini plugin
Parsed /Users/charendt/git/ankibinn/inventory.config inventory source with constructed plugin
@all:
  |--@ungrouped:
  |  |-----
  |  |--host1
```

If I change strict to yes in inventory.config I get the error `entry: 'mylocation' is undefined`:
```
ansible-inventory -i simpleinventory.ini -i inventory.config --graph -vvvv
ansible-inventory 2.7.9
  config file = /Users/charendt/git/ankibinn/ansible.cfg
  configured module search path = ['/Users/charendt/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible-inventory
  python version = 3.7.2 (default, Feb 12 2019, 08:15:36) [Clang 10.0.0 (clang-1000.11.45.5)]
Using /Users/charendt/git/ankibinn/ansible.cfg as config file
setting up inventory plugins
Parsed /Users/charendt/git/ankibinn/simpleinventory.ini inventory source with ini plugin
 [WARNING]:  * Failed to parse /Users/charendt/git/ankibinn/inventory.config with ini plugin: /Users/charendt/git/ankibinn/inventory.config:2: Expected key=value host variable
assignment, got: constructed

  File "/usr/local/lib/python3.7/site-packages/ansible/inventory/manager.py", line 275, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/local/lib/python3.7/site-packages/ansible/plugins/inventory/ini.py", line 134, in parse
    raise AnsibleParserError(e)

 [WARNING]:  * Failed to parse /Users/charendt/git/ankibinn/inventory.config with constructed plugin: failed to parse /Users/charendt/git/ankibinn/inventory.config: Could not
generate group from mylocation entry: 'mylocation' is undefined

  File "/usr/local/lib/python3.7/site-packages/ansible/inventory/manager.py", line 275, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/local/lib/python3.7/site-packages/ansible/plugins/inventory/constructed.py", line 132, in parse
    raise AnsibleParserError("failed to parse %s: %s " % (to_native(path), to_native(e)))

 [WARNING]: Unable to parse /Users/charendt/git/ankibinn/inventory.config as an inventory source

@all:
  |--@ungrouped:
  |  |-----
  |  |--host1
```

To show the hostvar is set I've listed the inventory:
```
ansible-inventory -i simpleinventory.ini -i inventory.config --list
{
    "_meta": {
        "hostvars": {
            "---": {},
            "host1": {
                "mylocation": "box1"
            }
        }
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    },
    "ungrouped": {
        "hosts": [
            "---",
            "host1"
        ]
    }
}
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by bcoca** [comment]
You can use  hostvars, just not  `host_vars/` nor `group_vars/` as those are not loaded at the point of inventory source evaluation, only after.

**Comment by s11-charendt** [comment]
Thanks for clarification, what would be a good approach to build groups based on host_vars/ ?

**Comment by bcoca** [comment]
group_by action

**Comment by s11-charendt** [comment]
Thanks for the hint, but I want to use the generated groups with --limit, so the groups must be assigned before execution of the playbook.

In the meantime, I found another issue, addressing the root cause of my issue: #44382

There is a workaround in #44382 which involves doing the host_vars/ inclusion by hand, so this would be executed twice. First, to generate Groups, and then nativ by ansible, which will work for now, but is suboptimal.

**Comment by bcoca** [comment]
host_vars/group_vars were never designed to be input to inventory, but a complement to inventory sources

**Comment by s11-charendt** [comment]
Ok, I've found a workarround, which requires minimal code. It's not pretty, but works for me. I've posted it in the root cause issue #44382
